### PR TITLE
Add transformations to polygons and points bounding box query (#143)

### PR DESCRIPTION
**Pull request by LucaMarconato**
Making raster queries compatible with coordinate systems.

### CODE REVIEW

**Comment by LucaMarconato** [comment]
TLDR; initially I wanted to ask an opinion to how to proceed, but I figured out the strategy will writing this text, so I leave it as documentation to understand all the cases of spatial queries with raster types.

Implementing the bounding box query for raster types and the support for coordinate systems is trickier than I thought. There are a few choices that I had to make and in this document I write down in details the strategy that I will implement. 

# Objective
The API for the bonding box query, currently implemented for points and shapes, looks like this:
```python=
@singledispatch
def bounding_box_query(
    element: Union[SpatialElement, SpatialData],
    axes: tuple[str, ...],
    min_coordinate: ArrayLike,
    max_coordinate: ArrayLike,
    target_coordinate_system: str,
) -> Optional[Union[SpatialElement, SpatialData]]:
    pass
```
When `element` is of type `SpatialData`, the method simply dispatches the call to each element and then puts the result together in a new `SpatialData` object, which is returned. So let's consider the case in which `element` is specifying points, shapes, images or labels. 

These are the other parameters of `bounding_box_query()`:
* `min_coordinate`, `max_coordinate`, `axes`, together they specify a 2D/3D axis-aligned bounding box;
* `target_coordinate_system`: is the name of the extrinsic coordinate systems in which the bounding box is specified. In other words `element.attrs["transform"][target_coordinate_system]` is a transformation that maps `element` to the coordinate system in which the bounding box is specified.

The function returns an element, of the same type as `element`, which is cropped in the intrinsic coordinate system and has a transformation that maps it back to `target_coordinate_system`. 

## The strategy with raster types
A working implementation for points and shapes can be found at this commit: c142295145f8fc0a6b5a3c653e7f4e96b7f5f176.

Let's see the raster case.

### Assumptions and shortcuts
Assumptions:
* As per schema, we assume that the images are c(z)yx, and the labels are (z)yx.

Shortcuts:
* The labels case is analogous to the image case, with just two points of care: 1) in the case of images we keep also all the c dimension; 2) in the case of labels if we performn a resampling/transformation of the data we need to disable interpolation otherwise we alter the values of the labels. This is done for instance also [here](https://github.com/scverse/spatialdata/blob/38c661c75be8ed76f6bc2906eb5c4664ce43dbc1/spatialdata/_core/_transform_elements.py#L269). 
* The multiscale case simplifies to the non-multiscale case. We just use the same code for each scale and we assemble the object back together, as done [here](https://github.com/scverse/spatialdata/blob/38c661c75be8ed76f6bc2906eb5c4664ce43dbc1/spatialdata/_core/_transform_elements.py#L294).

Therefore, in the following we will focus on non-multiscale images.

### All the cases and the corresponding query results
I anticipate that my plan for the code is to return the answer as a single call of `dask_image.ndinterp.affine_transform`.

TLDR; skip to the recap.

To visualize which data we would get from a query it is useful to consider all the cases we operate with. Let's consider all the 8 combinations of the following:
1. the data is 2D or 3D;
2. the data, transformed to `target_coordinate_system`, has *dimension* 2D or 3D (e.g. a 3D image after an identity is a 3D image, but after a projection to a 2D plane is a 2D image; a 2D image embedded in a 3D space has still 2 dimensions);
3. the data, transformed to `target_coordinate_system` has 2D or 3D *coordinates* (e.g. a 2D image after an identity has 2D coordiantes, but could also be transformed with an affine matrix that embeds it in the 3D space, so having 3D coordinates).

To compute the *dimension* and number of *coordinates* of the transformed data we can
1. get a matrix from the function 
```_get_affine_for_element(element: SpatialElement, transformation: BaseTransformation) -> Affine```
which automatically computes the affine transformations with axis equal to the ones of the intrinsic coordinate system of the element, and output axes inferred calling 
```_get_current_output_axes(transformation: BaseTransformation, input_axes: tuple[ValidAxis_t, ...]) -> tuple[ValidAxis_t, ...]```
2. remove the eventual `c` dim from input and output axes (this may remove 1 column and/or 1 row from the matrix)
3. compute the rank -> this is the *dimension* of the transformed data;
4. compute `.shape[0]` -> this is the number of *coordinates* of the transformed data.

In theory there could be 8 combinations, but only 5 are possible.

These cases are never possible:
* 2D image with transformed *dimension* being 3D.
* image with transfomed *dimension* being 3D but transformed *coordinates* being 2D.

5 remaining cases are possible:
1. 2D image with transformed *dimension* 2D and transformed *coordinates* 2D;
2. 2D image with transformed *dimension* 2D and transformed *coordinates* 3D;
* ~~2D image with transformed *dimension* 3D and transformed *coordinates* 2D;~~
* ~~2D image with transformed *dimension* 3D and transformed *coordinates* 3D;~~
3. 3D image with transformed *dimension* 2D and transformed *coordinates* 2D;
4. 3D image with transformed *dimension* 2D and transformed *coordinates* 3D;
* ~~3D image with transformed *dimension* 3D and transformed *coordinates* 2D;~~
5. 3D image with transformed *dimension* 3D and transformed *coordinates* 3D.

Let's visualize the following cases:

#### Query result from cases 1, 5
##### Subcase: non-rotation
The bounding box, transformed to the intrinsic coordinate system, is still a rectangular (2D) or parallelepiped (3D) bounding box, with axes aligned to the data shape.



If the bounding box is bigger than the data, we have two possibilities on what to return:
a. add a padding (with zeros);
b. return just the data inside the boundning box.



Case b is appealing, but I prefer case a, because it makes the implementation way easier.

##### Subcase: anysotropic transformation
Anysotropic transformations that are not covered by the previous case are rotations. The bounding box, transformed to the intrinsic coordinate system, is not a rotated bounding box. The sides of the bounding box are orthogonal between each other, but are not aligned with the sides of the data. 

Here we have to make an *additional* choice:
a. we return the result as an axis aligned crop
b. we return the result keeping the original image orientation


In the case of b. we have a few more sub choices:
b1. we "pad the corners" with zeros
b2. we keep the corners with image data


Case a is appealing, but I prefer case b2, I'll explain later why.
##### Subcase: non-anisotroic transformation
Here the bounding box, transformed to the intrinsic coordinate system, is a quadrilateral/parallelepieped with axes that are not orthogonal between each other.


The cases b. and c. from the subcase before are still possible. Case a. not anymore, unless we "stretch" the quadrilateral to fit a rectangle.


Case a starts be less appealing; I still prefer case c.
#### Query result from case 3
This is analogous to the cases 1 and 5, with the difference that the affine transformation is not invertible (it's not injective). The returned value is always a volume. The choices we had to make before are also valid for this case.

#### Query result from case 4
This case will be analogous to case 2, using the same logic as for case 3.

#### Query result from case 2
Here the image in `target_coordinate_system` lies on a 2D plane, while the query area is a 3D bounding box. The intersection of the bounding box with a plane is a 2D polygon that can have 3, 4, 5 or 6 verticies. In the figure the queried region is an hexagon.


Again, cases b1 and b2 are possible, but cases a is not possible anymore. 



# Recap (and the reason for case b2)
The last example is the reason why I exclude case a before.
The reason why I prefer the case b1 over the case b2, is that it is easier to implement. I can implement the query as a lazy operation that is carried on as a single call of the function
```
transformed_dask = dask_image.ndinterp.affine_transform(
    data,
    matrix=matrix,
    offset=offset,
    output_shape=output_shape,
    **kwargs,
)
```
`matrix` is the affine matrix, and doesn't need to be inverted.

The `output_shape` and `offset` are easy to compute for the cases 1, 2, 5.
They are simply respectively the size and the origin of the bounding box, in the intrinsic coordinate system, of the inverse-transformed bounding box from the `target_coordinate_system`.

In the cases 3, 4 the matrix is not invertible, but one can play around with the linear system determined by the matrix (which has solutions with degree of freedom 1) to get a 2D bounding box on the plane perpendicular to a vector generating the infinite solutions, and finally compute the bounding limits on the last direction. Something like that.

## Further points
I propose to return, together with the queried data, an optional rasterized mask of how the bounding box transforms to the intrinsic coordinate system.

**Comment by LucaMarconato** [comment]
Ready for review @kevinyamauchi. I have implemented only the cases 1 and 5 (put the wrong commit message in the second last commit). I would implement cases 2, 3, 4 after the paper since we won't use them anytime soon.

**Comment by LucaMarconato** [comment]
Forgot to add the multiscale case, doing it now.

**Comment by LucaMarconato** [comment]
@kevinyamauchi [I added an example (with a napari viz included)](https://github.com/scverse/spatialdata/blob/65cc2f154ca2d452fdabfad137f80e9c9671b3b2/tests/_core/test_spatial_query.py#LL364) that showcases the bonding box query. Please follow the docstring as there I explain how to get the most out of the example. 

As the example shows, this PR can be used for the tiling for deep learning in two ways:
1) our tiles for deep learning are in the intrinsic coordinate system. Here, if the data in the target coordinate system is rotated, the queried data will contain also data in the surrounding of the query, as explained.
2) when more control is needed, one can get the exact crop location in the target coordinate system. Nevertheless the procedure for doing this (showed in the example) is not pixel perfect and is a quite complex. We need to implement a way for cropping the data, directly in the target_coordinate_system and returning it in the target coordinate system.

# The new method
We need to discuss how to include this new method in the APIs (we should make the requests more powerful). For the moment I will implement the method under another name: `_resample_raster()`. Later we can decide how to include it as a subroutine of the bounding box query APIs.

This will be the signature:
```
@bounding_box_query.register(MultiscaleSpatialImage)
def _(
    image: Union[SpatialImage|MultiscaleSpatialImage],
    axes: tuple[str, ...],
    min_coordinate: ArrayLike,
    max_coordinate: ArrayLike,
    target_coordinate_system: str,
    target_unit_to_pixels: Optional[float],
    target_width: Optional[float],
    target_height: Optional[float],
    target_depth: Optional[float],
) -> SpatialImage:
```
The method:
* will always return a `SpatialImage` of the desired resolution. If the input is a multiscale, for doing this it will use the closest (larger if available) level of the multiscale.
* The desired resolution is specified by one (and only one of the 4 optional parameters in the end), and it will have the same axes ratio of the specified axis-aligned bounding box
* If the bounding box doesn't fully overlap with the data, it will be filled with zeros. The bounding box can even not overlap at all, and it will just return a tensor of zeros.

We can refine the new APIs on the go, for the moment I will implement this structure.

**Comment by LucaMarconato** [comment]
Here is the implementation of the new function. It is not usable yet (there is one bug and needs some small fixes), but you it will be more or less this: https://github.com/scverse/spatialdata/blob/866d1c0bb31b6b3c2a8b60624a667064a4a49ad5/spatialdata/_core/_rasterize.py#L154

**Review by kevinyamauchi — APPROVED** [review]
Great work, @LucaMarconato ! I think this is good to go. Please see my minor comments below and feel free to merge when ready.

File: spatialdata/_core/_spatial_query.py [hunk 1]
```diff
@@ -57,154 +187,365 @@ def __post_init__(self) -> None:
...
+        max_coordinate=max_coordinate,
+    )
     from spatialdata._core._spatialdata_ops import (
         get_transformation,
         set_transformation,
     )
 
+    # get the transformation from the element's intrinsic coordinate system to the query coordinate space
+    transform_to_query_space = get_transformation(image, to_coordinate_system=target_coordinate_system)
+    assert isinstance(transform_to_query_space, BaseTransformation)
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply spatialdata/_core/_spatial_query.py:323]: added an exception. It will not print the element name (since it is not known from the variables in the scope), but it will print the element. This exception was already present in `get_transformation()`, now it is also printing the element.

File: spatialdata/_core/_spatial_query.py [hunk 2]
```diff
@@ -57,154 +187,365 @@ def __post_init__(self) -> None:
...
+            raise ValueError(
+                f"Invalid case. The bounding box axes are {axes}, the spatial axes in {target_coordinate_system} are {output_axes_without_c}"
+            )
+    spatial_transform = Affine(m_without_c, input_axes=input_axes_without_c, output_axes=output_axes_without_c)
+    spatial_transform_bb_axes = Affine(
+        spatial_transform.to_affine_matrix(input_axes=input_axes_without_c, output_axes=axes),
+        input_axes=input_axes_without_c,
+        output_axes=axes,
+    )
+    if case in [1, 2, 5]:
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply spatialdata/_core/_spatial_query.py:371]: only computing `bounding_box_corners` is different from case 1+5 and case 2. The second part of the code is shared. But now I will simplify since we don't support case 3 and 4 anymore (as discussed via chat).

File: spatialdata/_core/_spatial_query.py [hunk 3]
```diff
@@ -57,154 +187,365 @@ def __post_init__(self) -> None:
...
+            bounding_box_corners = get_bounding_box_corners(
+                min_coordinate=min_coordinate,
+                max_coordinate=max_coordinate,
+                axes=axes,
+            )
+        else:
+            assert case == 2
+            # TODO: we need to intersect the plane in the extrinsic coordiante system with the 3D bounding box. The
+            #  vertices of this polygons needs to be transformed to the intrinsic coordinate system
+            raise NotImplementedError("Case 2 is not implemented yet")
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply spatialdata/_core/_spatial_query.py:382]: done

File: spatialdata/_core/_spatial_query.py [hunk 4]
```diff
@@ -57,154 +187,365 @@ def __post_init__(self) -> None:
...
+        translation = inverse.matrix[0:-1, -1]
+
+        intrinsic_bounding_box_corners = DataArray(
+            bounding_box_corners.data @ rotation_matrix.T + translation,
+            coords={"corner": range(len(bounding_box_corners)), "axis": list(inverse.output_axes)},
+        )
+    else:
+        assert case in [3, 4]
+        # TODO: we need to work with the underspecified linear system of equations, etc..
+        raise NotImplementedError("Case 3 and 4 are not implemented yet")
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply spatialdata/_core/_spatial_query.py:395]: done (a bit above in the code)

File: spatialdata/_core/_spatial_query.py [hunk 5]
```diff
@@ -57,154 +187,365 @@ def __post_init__(self) -> None:
...
+            if 0 in xdata.shape:
+                if k == "scale0":
+                    return None
+            else:
+                d[k] = xdata
+        query_result = MultiscaleSpatialImage.from_dict(d)
+    query_result = compute_coordinates(query_result)
+
+    if not np.allclose(np.array(translation_vector), 0):
+        translation_transform = Translation(translation=translation_vector, axes=axes)
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply spatialdata/_core/_spatial_query.py:439]: done, pasting it here:
```
  # the bounding box, mapped back to the intrinsic coordinate system is a set of points. The bounding box of these
  # points is likely starting away from the origin (this is described by translation_vector), so we need to prepend
  # this translation to every transformation in the new queries elements (unless the translation_vector is zero,
  # in that case the translation is not needed)
```

File: tests/_core/test_spatial_query.py [hunk 6]
```diff
@@ -213,10 +322,140 @@ def test_bounding_box_circles():
...
+    for element in result._gen_elements_values():
+        d = get_transformation(element, get_all=True)
+        new_d = {k.replace("global", "cropped"): v for k, v in d.items()}
+        set_transformation(element, new_d, set_all=True)
+
+    VISUALIZE = False
+    if VISUALIZE:
+        from napari_spatialdata import Interactive
+
+        Interactive([full_sdata, result])
```
[the remark this answers is not in the document]
Comment (LucaMarconato) [reply tests/_core/test_spatial_query.py:361]: done thanks
